dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/AsyncRpcResult.java

```diff
@@ -25,176 +25,174 @@
+        return getAppResponse().getValue();
     }
 
     @Override
-    public Throwable getException() {
-        return getRpcResult().getException();
+    public void setValue(Object value) {
+        AppResponse appResponse = new AppResponse();
+        appResponse.setValue(value);
+        this.complete(appResponse);
     }
 
     @Override
-    public boolean hasException() {
-        return getRpcResult().hasException();
+    public Throwable getException() {
+        return getAppResponse().getException();
     }
 
     @Override
-    public Object getResult() {
-        return getRpcResult().getResult();
-    }
-
-    public CompletableFuture getValueFuture() {
-        return valueFuture;
+    public void setException(Throwable t) {
+        AppResponse appResponse = new AppResponse();
+        appResponse.setException(t);
+        this.complete(appResponse);
     }
 
-    public CompletableFuture<Result> getResultFuture() {
-        return resultFuture;
-    }
-
-    public void setResultFuture(CompletableFuture<Result> resultFuture) {
-        this.resultFuture = resultFuture;
+    @Override
+    public boolean hasException() {
+        return getAppResponse().hasException();
     }
 
-    public Result getRpcResult() {
+    public Result getAppResponse() {
         try {
-            if (resultFuture.isDone()) {
-                return resultFuture.get();
+            if (this.isDone()) {
+                return this.get();
             }
         } catch (Exception e) {
             // This should never happen;
             logger.error("Got exception when trying to fetch the underlying result from AsyncRpcResult.", e);
         }
-        return new RpcResult();
+        return new AppResponse();
     }
 
     @Override
     public Object recreate() throws Throwable {
-        return valueFuture;
+        RpcInvocation rpcInvocation = (RpcInvocation) invocation;
+        if (InvokeMode.FUTURE == rpcInvocation.getInvokeMode()) {
+            AppResponse appResponse = new AppResponse();
+            CompletableFuture<Object> future = new CompletableFuture<>();
+            appResponse.setValue(future);
+            this.whenComplete((result, t) -> {
+                if (t != null) {
+                    if (t instanceof CompletionException) {
+                        t = t.getCause();
+                    }
+                    future.completeExceptionally(t);
+                } else {
+                    if (result.hasException()) {
+                        future.completeExceptionally(result.getException());
+                    } else {
+                        future.complete(result.getValue());
+                    }
+                }
+            });
+            return appResponse.recreate();
+        } else if (this.isDone()) {
+            return this.get().recreate();
+        }
+        return (new AppResponse()).recreate();
+    }
+
+    public Result thenApplyWithContext(Function<Result, Result> fn) {
```

Comment:
TBD.
I will review it later.

Reply:
Need to review the usage and learn the completableFuture.

Reply:
In a recent perf test, I found this place costs quite a percentage of CPU time.